File: src/content/docs/topics/cs/coding-problems/trees/098-validate-binary-search-tree-approach2.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func isValidBST(root *TreeNode) bool {
	prev := -1 << 62 // sentinel for "no previous"
	hasPrev := false

	var inorder func(node *TreeNode) bool
	inorder = func(node *TreeNode) bool {
		if node == nil {
			return true
		}
		if !inorder(node.Left) {
			return false
		}
		if hasPrev && node.Val <= prev {
			return false
		}
		prev = node.Val
		hasPrev = true
		return inorder(node.Right)
	}
	return inorder(root)
}

func runTests() {
	assert(isValidBST(&TreeNode{Val: 2, Left: &TreeNode{Val: 1}, Right: &TreeNode{Val: 3}}) == true)

	assert(isValidBST(&TreeNode{Val: 5,
		Left:  &TreeNode{Val: 1},
		Right: &TreeNode{Val: 4, Left: &TreeNode{Val: 3}, Right: &TreeNode{Val: 6}},
	}) == false)

	assert(isValidBST(&TreeNode{Val: 1}) == true)
	assert(isValidBST(nil) == true)

	assert(isValidBST(&TreeNode{Val: 3,
		Left:  &TreeNode{Val: 1, Left: &TreeNode{Val: 0}, Right: &TreeNode{Val: 2}},
		Right: &TreeNode{Val: 5, Left: &TreeNode{Val: 4}, Right: &TreeNode{Val: 6}},
	}) == true)

	assert(isValidBST(&TreeNode{Val: 5,
		Left:  &TreeNode{Val: 4},
		Right: &TreeNode{Val: 6, Left: &TreeNode{Val: 3}, Right: &TreeNode{Val: 7}},
	}) == false)

	fmt.Println("all tests pass")
}

func main() { runTests() }

File: src/content/docs/topics/cs/coding-problems/trees/098-validate-binary-search-tree-approach2.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

class TreeNode {
    val: number;
    left: TreeNode | null;
    right: TreeNode | null;
    constructor(val: number = 0, left: TreeNode | null = null, right: TreeNode | null = null) {
        this.val = val; this.left = left; this.right = right;
    }
}

function buildTree(vals: (number | null)[]): TreeNode | null {
    if (!vals.length) return null;
    const root = new TreeNode(vals[0] as number);
    const q: TreeNode[] = [root];
    let i = 1;
    while (q.length && i < vals.length) {
        const node = q.shift()!;
        if (i < vals.length && vals[i] !== null) { node.left = new TreeNode(vals[i] as number); q.push(node.left); }
        i++;
        if (i < vals.length && vals[i] !== null) { node.right = new TreeNode(vals[i] as number); q.push(node.right); }
        i++;
    }
    return root;
}

function isValidBST(root: TreeNode | null): boolean {
    let prev = -Infinity;

    function inorder(node: TreeNode | null): boolean {
        if (!node) return true;
        if (!inorder(node.left)) return false;   // L2: recurse left, O(h) stack depth
        if (node.val <= prev) return false;       // L3: O(1) monotonicity check
        prev = node.val;                          // L4: O(1) update
        return inorder(node.right);              // L5: recurse right
    }

    return inorder(root);
}

assert(isValidBST(buildTree([2, 1, 3])) === true);
assert(isValidBST(buildTree([5, 1, 4, null, null, 3, 6])) === false);
assert(isValidBST(buildTree([1])) === true);
assert(isValidBST(null) === true);
assert(isValidBST(buildTree([3, 1, 5, 0, 2, 4, 6])) === true);
assert(isValidBST(buildTree([5, 4, 6, null, null, 3, 7])) === false);
console.log("all tests pass");
